rfcs/45-stageless.md

```diff
@@ -0,0 +1,727 @@
+    .add_system(flush_commands.in_set("DespawningFlush").after(CommandFlush::PostUpdate).before(CommandFlush::EndOfFrame))
+    .add_system(fork_projectiles.after("DespawningFlush"));
+  }
+}
+```
+
+### Complex control flow
+
+Occasionally, you may find yourself yearning for more complex system control flow than "every system runs once in a loop".
+When that happens: **create an exclusive system and run a schedule in it.**
+
+Within an exclusive system, you can freely fetch the desired schedule from the `App` with the `&Schedules` system parameter and use `Schedule::run(&mut world)`, applying each of the systems in that schedule a single time to the world of the exclusive system.
+However, because you're in an ordinary Rust function you're free to use whatever logic and control flow you desire: branch and loop in whatever convoluted fashion you need!
+
+This can be helpful when:
+
+- you want to run multiple steps of a set of game logic during a single frame (as you might see in complex turn-based games)
+- you need to branch the entire schedule to handle a particularly complex bit of game logic
+- you need to integrate a complex external service, like scripting or a web server
+- you want to switch the executor used for some portion of your systems
+
+If you need absolute control over the entire schedule, consider having a single root-level exclusive system.
+
+#### Fixed time steps
+
+Running a set of systems (typically physics) according to a fixed time step is a particularly critical case of this complex control flow.
+These systems should run a fixed number of times for each second elapsed.
+The number of times these systems should be run varies (it may be 0, 1 or more each frame, depending on the time that the frame took to complete).
+
+Simply adding run criteria is inadequate: run criteria can only cause our systems to run a single time, or not run at all.
+By moving this logic into its own schedule within an exclusive system, we can loop until the accumulated time has been spent, running the schedule repeatedly and ensuring that our physics always uses the same delta-time and stays synced with the clock, even in the face of serious stutters.
+
+## Implementation strategy
+
+Let's take a look at what implementing this would take:
+
+1. Completely rip out:
+   1. `Stage`
+   2. `Schedule`
+   3. `SystemDescriptor`
+   4. `IntoExclusiveSystem`
+   5. `SystemContainer` and friends
+2. Build out a basic schedule abstraction:
+   1. Create a `ScheduleLabel` trait
+   2. Store multiple schedules in a keyed `SystemRegistry` resource
+   3. Support adding systems to the schedule
+      1. These should be initialized by definition: see [PR #2777](https://github.com/bevyengine/bevy/pull/2777)
+   4. Support adding systems to specific schedules and create startup system API
+3. Rebuild core command processing machinery
+   1. Begin with trivial event-style commands
+   2. Explore how to regain parallelism

```

Comment:
Hand waving "explore how to regain parallelism" isn't satisfying. Parallel command composition is a core piece of the puzzle. A single "simple" `flush_commands` system is interesting and powerful, but deciding how to efficiently produce and apply commands in parallel is one of the "hard problems" in stageless. "Begin with trivial event-style commands" (if using Bevy events) creates a hard sync point and is therefore untenable. Using a shared crossbeam event channel will incur per-command overhead that might not be worth it, and it would throw away the efficient dyn-stack impl we have today. Naively it would require per-command-boxing, which we already know is significantly worse than the dyn-stack.